### PR TITLE
Plfm 2424

### PR DESCRIPTION
**Pull request by brucehoff**

### CODE REVIEW

File: client/synapseJavaClient/src/main/java/org/sagebionetworks/client/SynapseClientImpl.java [hunk 1]
```diff
@@ -3366,6 +3366,105 @@ public MessageToUser sendMessage(MessageToUser message) throws SynapseException
...
 	}
 	
+
+	/**
+	 * uploads a String to S3 using the chunked file upload service
+	 * Note:  Strings in memory should not be large, so we limit to the size of one 'chunk'
+	 */
+	@Override
+    public String uploadToFileHandle(String content, String contentType) throws SynapseException {
+    	if (content==null || content.length()==0) throw new IllegalArgumentException("Missing content.");
```
Comment (marcel-blonk) [inline client/synapseJavaClient/src/main/java/org/sagebionetworks/client/SynapseClientImpl.java]: Empty emails should be possible?

File: client/synapseJavaClient/src/main/java/org/sagebionetworks/client/SynapseClientImpl.java [hunk 2]
```diff
@@ -3366,6 +3366,106 @@ public MessageToUser sendMessage(MessageToUser message) throws SynapseException
...
+		CompleteAllChunksRequest cacr = new CompleteAllChunksRequest();
+		cacr.setChunkedFileToken(token);
+		cacr.setChunkNumbers(chunkNumbers);
+		UploadDaemonStatus status = startUploadDeamon(cacr);
+		State state = status.getState();
+		if (state.equals(State.FAILED)) throw new IllegalStateException("Message creation failed: "+status.getErrorMessage());
+			
+		long backOffMillis = 100L; // initially just 1/10 sec, but will exponentially increase
+		int backOffCounter = 0;
+		while (state.equals(State.PROCESSING) && backOffCounter++<MAX_BACKOFF_TRIES) {
```
Comment (marcel-blonk) [inline client/synapseJavaClient/src/main/java/org/sagebionetworks/client/SynapseClientImpl.java]: 100 seconds max might be too small

File: client/synapseJavaClient/src/main/java/org/sagebionetworks/client/SynapseClientImpl.java [hunk 3]
```diff
@@ -3366,6 +3366,106 @@ public MessageToUser sendMessage(MessageToUser message) throws SynapseException
...
+			} catch (InterruptedException e) {
+				// continue
+			}
+			status = getCompleteUploadDaemonStatus(status.getDaemonId());
+			state = status.getState();
+			if (state.equals(State.FAILED)) throw new IllegalStateException("Message creation failed: "+status.getErrorMessage());
+			backOffMillis *= 2; // exponential backoff
+		}
+		
+		if (state.equals(State.FAILED)) throw new IllegalStateException("Message creation failed: "+status.getErrorMessage());
```
Comment (marcel-blonk) [inline client/synapseJavaClient/src/main/java/org/sagebionetworks/client/SynapseClientImpl.java]: Bruce identified bug: need to say !success here

File: lib/communicationUtilities/src/main/java/org/sagebionetworks/utils/MD5ChecksumHelper.java [hunk 4]
```diff
@@ -39,6 +40,17 @@ public static String getMD5Checksum(File file) throws IOException {
 		byte[] b = createChecksum(new FileInputStream(file));
 		return getHexString(b);
 	}
+	
+	public static String getMD5ChecksumForString(String content) throws IOException {
+		ByteArrayInputStream inputStream = null;
+		try {
+			inputStream = new ByteArrayInputStream(content.getBytes());
```
Comment (marcel-blonk) [inline lib/communicationUtilities/src/main/java/org/sagebionetworks/utils/MD5ChecksumHelper.java]: Specify encoding

File: client/synapseJavaClient/src/main/java/org/sagebionetworks/client/SynapseClientImpl.java [hunk 5]
```diff
@@ -3389,6 +3389,108 @@ public MessageToUser sendMessage(MessageToUser message) throws SynapseException
 		}
 	}
 	
+
+	/**
+	 * uploads a String to S3 using the chunked file upload service
+	 * Note:  Strings in memory should not be large, so we limit to the size of one 'chunk'
+	 */
+	@Override
+    public String uploadToFileHandle(String contentString, String contentType) throws SynapseException {
```
Comment (kaysoky) [inline client/synapseJavaClient/src/main/java/org/sagebionetworks/client/SynapseClientImpl.java:3398]: It seems like this method and createFileHandle have a lot of logic in common.  If possible, we should combine them.

File: client/synapseJavaClient/src/main/java/org/sagebionetworks/client/SynapseClientImpl.java [hunk 6]
```diff
@@ -3389,6 +3389,108 @@ public MessageToUser sendMessage(MessageToUser message) throws SynapseException
...
+    	if (contentString==null || contentString.length()==0) throw new IllegalArgumentException("Missing content.");
+    	
+    	// Note, we make sure to convert from String to byte[] just once.
+		byte[] content = contentString.getBytes();
```
Comment (marcel-blonk) [inline client/synapseJavaClient/src/main/java/org/sagebionetworks/client/SynapseClientImpl.java:3402]: Add specific encoding

**Pull request merged**